### PR TITLE
[WIP] implementation compatible with libyang2

### PR DESCRIPTION
**Pull request by Dajvid**
Hi, as I promised some time ago, I'm opening a pull request with my current progress on the python wrapper for libyang2. It still needs a lot of love, but I would like to get some feedback.

As I was writing it while libyang2 was still under heavy development without stable API, and I wasn't even able to keep up with the latest changes lately (this version is compatible with libyang2 on commit 45802b59e0ef479a1af8cf373bb24b50d348bbbd), I will have to do a lot of refactoring.

Overall I'm quite happy with context and schema modules, on the other hand, I'm not happy with the data module, I found it very hard to create friendly API for some complicated lyd_* functions. I tried some approaches with generalization and merging of those functions, but without much success, so I'm planning to try to implement the equivalent of your "data dictionary" for that.

I'm also planning to:
- Reuse your setup for CI/CD and installation.
- Work on backwards compatibility (methods vs properties, add methods to ensure compatibility when needed and possible...).
- Try to reuse most of your e2e tests.
- Write proper documentation.

### CODE REVIEW

**Comment by rjarry** [comment]
Hi @Dajvid 

I'll take some time to make a proper review this week.

Thanks

**Comment by rjarry** [comment]
Hi David,

Before digging deeper into the code I have a few general remarks:

- Please preserve the existing MIT license.
- Do not remove my copyrights, add a line with your name.
- Do not change code unless it is necessary (`README.rst`, `cffi/build.py`, etc.).

Also, I wonder if it could be possible to continue to support both versions of libyang.

**Review by rjarry — CHANGES_REQUESTED** [review]
Hi David,

Before digging deeper into the code I have a few general remarks:

- Please preserve the existing MIT license.
- Do not remove my copyrights, add a line with your name.
- Do not change code unless it is necessary (`README.rst`, `cffi/build.py`, etc.).
- Since you are removing the `libyang` package. No need to use a different `libyang2`.

Also, I wonder if it could be possible to continue to support both versions of libyang.

**Comment by wisotzky** [comment]
@rjarry, I don't think there is any value to support both version. Since libyang2 was released a month ago, the original libyang was obsoleted. As long the python API stays compatible, can move to version 2 straight ahead.

@Dajvid, it would be great to get some timelines... There are some important enhancements in libyang2, which I would plan to use. Not sure, if I can support here...

**Comment by Dajvid** [comment]
I'm sorry, didn't have time for this lately, but I can get back to work now and during the summer.

I would definitely appreciate some help, mainly with the improvements of data module, overall api improvements, installation scripts and CI setup.

Support both versions at once would be very complicated and I don't think it's a good idea.
Full compatibility across both versions is not possible in my opinion, but both versions can be similar and require just small changes. Similarly as libyang2 is not fully backwards compatible...

@rjarry I will adress your remarks until monday and then we can continue.
@wisotzky It should be already somewhat usable (obvisously with a lot of complications, but main parts are covered), I need to do a lot of refactoring because this was designed and implemented mostly during last year before libyang2 was released and didn't have stable api yet at that time. And as for the timeline, I would definitely like to get this done and release ready until the end of this summer (or end of September to be precise), because after that my free time will be very limited again.

File: .gitignore [hunk 1]
```diff
@@ -1,10 +1,138 @@
...
+dmypy.json
+
+# Pyre type checker
+.pyre/
+
+# pytype static type analyzer
+.pytype/
+
+# Cython debug symbols
+cython_debug/
```
Comment (rjarry) [inline .gitignore:138]: I am not sure this is relevant to libyang2. Please only add the required changes. We do not need to exclude all this.

File: MANIFEST.in [hunk 2]
```diff
@@ -1,10 +0,0 @@
-include cffi/build.py
-include cffi/cdefs.h
-include cffi/source.c
-include libyang/*.py
-include libyang/py.typed
-include libyang/VERSION
-include setup.py
-include setup.cfg
-include LICENSE
-include README.rst
```
Comment (rjarry) [inline MANIFEST.in:10]: Why remove this file?

File: Makefile [hunk 3]
```diff
@@ -1,15 +1,14 @@
-# Copyright (c) 2018-2020 Robin Jarry
```
Comment (rjarry) [inline Makefile:1]: Please preserve my copyritghts

File: Makefile [hunk 4]
```diff
@@ -1,15 +1,14 @@
...
+	python -m unittest
 
-format:
-	tox -e format
-
-.PHONY: lint tests format
+coverage:
+	python -m coverage run -m unittest discover -c tests/
+	python -m coverage report
+	python -m coverage html
```
Comment (rjarry) [inline Makefile:14]: I do not think you need to change anything in this file.

File: README.rst [hunk 5]
```diff
@@ -1,273 +0,0 @@
...
-   request.
-
-#. Wait for a reviewer to merge your work. If minor adjustments are requested,
-   use ``git commit --fixup $sha1`` to make it obvious what commit you are
-   adjusting. If bigger changes are needed, make them in new separate commits.
-   Once the reviewer is happy, please use ``git rebase --autosquash`` to amend
-   the commits with their small fixups (if any), and ``git push --force`` on
-   your topic branch.
-
-Thank you in advance for your contributions!
```
Comment (rjarry) [inline README.rst:273]: Why delete this file?

File: cffi/build.py [hunk 6]
```diff
@@ -30,17 +31,16 @@ def search_paths(env_var: str) -> List[str]:
...
-        py_limited_api=False,
+with open(os.path.join(HERE, 'source.c')) as f:
+    ffibuilder.set_source('_libyang',
+                          f.read(),
+                          libraries=['yang'],
+                          extra_compile_args=EXTRA_CFLAGS,
+                          extra_link_args=EXTRA_LDFLAGS,
+                          include_dirs=HEADERS,
+                          library_dirs=LIBRARIES,
+                          py_limited_api=False,
```
Comment (rjarry) [inline cffi/build.py:42]: Please do not change lines unless required. Do not mix reformatting/refactoring with libyang2 related stuff. Also, I doubt this matches black coding style :)

**Review by rjarry — CHANGES_REQUESTED** [review]
Hi @Dajvid 

I did not read everything. This is almost impossible to review. Too many changes. Too many of them are completely unrelated to libyang2.

Git history is very important (to me at least :)) and I wait to avoid changing code unless we have a good reason to do so. And more importantly, never mix topics in the same commits.

I think the best course of action would be to discuss the overall design choices that you have made and see how that can be integrated into the current code base without a complete rewrite.

Would that be ok for you?

**Pull request closed without merging**